patches/api/0449-add-number-format-api.patch

```diff
@@ -0,0 +1,185 @@
++     *
++     * @param style the style to apply on the number
++     * @return a styled number format
++     */
++    static StyledFormat styled(@NotNull Style style) {
++        return new StyledFormatImpl(style);
++    }
++
++    /*
++     * Creates a scoreboard number format that replaces the score number with a chat component.
++     *
++     * @param component the component to replace the number with
++     * @return a fixed number format
++     */
++    static FixedFormat fixed(@NotNull Component component) {
++        return new FixedFormatImpl(component);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormat.java b/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormat.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..3c644e150bed9df74e9168b2fcbbaab58cb6a654
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormat.java
+@@ -0,0 +1,23 @@
++package io.papermc.paper.scoreboard.numbers;
++
++import net.kyori.adventure.text.format.NamedTextColor;
++import net.kyori.adventure.text.format.Style;
++
++/*
++ * A scoreboard number format that applies a custom formatting to the score number.
++ */
++public interface StyledFormat extends NumberFormat {
++
++    StyledFormat NO_STYLE = NumberFormat.styled(Style.empty());
++
++    StyledFormat SIDEBAR_DEFAULT = NumberFormat.styled(Style.empty().color(NamedTextColor.RED));
++
++    StyledFormat PLAYER_LIST_DEFAULT = NumberFormat.styled(Style.empty().color(NamedTextColor.YELLOW));
++
++    /*
++     * The style that is being applied to the number in the score
++     * @return the style to apply
++     */
++    Style style();
++
++}
+diff --git a/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormatImpl.java b/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormatImpl.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..2489f63707bce74ecd8045948de73d529c06a446
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/scoreboard/numbers/StyledFormatImpl.java
+@@ -0,0 +1,5 @@
++package io.papermc.paper.scoreboard.numbers;
++
++import net.kyori.adventure.text.format.Style;
++
++record StyledFormatImpl(Style style) implements StyledFormat {}
+diff --git a/src/main/java/org/bukkit/scoreboard/Objective.java b/src/main/java/org/bukkit/scoreboard/Objective.java
+index a193ffabb05160b462dee1ba8f687fdbc84405b6..703b7788fbed7aff7f8274f40df3d4b6bfa30804 100644
+--- a/src/main/java/org/bukkit/scoreboard/Objective.java
++++ b/src/main/java/org/bukkit/scoreboard/Objective.java
+@@ -175,4 +175,20 @@ public interface Objective {
+      */
+     @NotNull Score getScoreFor(@NotNull org.bukkit.entity.Entity entity) throws IllegalArgumentException, IllegalStateException;
+     // Paper end - improve scoreboard entries
++
++    //Paper start - number format
++    /*
++     * Gets the number format for this objective's entries (scores)
++     * @return this objective's number format
```

Comment:
Might be worth to comment the current default of this value (just the number) in the javadocs.

Reply:
Might need more comments on the `@Nullable` vs `@NotNull` given we expose NO_STYLE in the API now.

Reply:
> Might be worth to comment the current default of this value (just the number) in the javadocs.

Well in server terms there is no default value as far as I can tell. That is fully up to the client what to style that like. It would be hard to describe how it would look like in game, since objectives and their scores can be more than just the scoreboard sidebar...

I would suggest a wording like that maybe?
For score (getter):
`The number format might be null if the score has not been set yet or when the objective default (if set) 
or the client default number format is used`
For score (setter):
`@param format the number format to set, pass null to reset format to the objective (if set) or to the client default`

Objective(getter): 
`@return this objective's number format or null if the client default is being used`
Objective(setter):
`@param format the number format to set, pass null to reset format to default`

Reply:
I think we stick with nullable getter/setter. I don't think we wanna make the assumption that noStyle StyledFormat will always be the same as the default format because @MenschenToaster is right, that is a client implementation detail.